### ISSUE TITLE
false negative when unboxing potentially null value in element accesses of arrays

### ISSUE THREAD

**Issue by theosotr** [typetools/checker-framework#6804]
## Command

```
javac \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
  -J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED \
  -J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED \
-processor org.checkerframework.checker.nullness.NullnessChecker \
-cp checker-framework-3.47.0/checker/dist/checker.jar Main.java
```
 ## File
 
 ```java
import org.checkerframework.checker.nullness.qual.*;


@FunctionalInterface
interface A<T> {
    T get();
}

class Main {
  public static void main(String[] args) {
      A<@Nullable Integer> y = () -> null;
      String x = (new String[]{"val"})[y.get()];
  }
}
```

## Actual behavior

The code passes the checks, but I get a NPE at runtime

```
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "A.get()" is null
        at Main.main(Main.java:12)
```
 
## Expected behavior

```
Main.java:12: error: [unboxing.of.nullable] unboxing a possibly-null reference y.get()
```

**Comment by smillst** [comment]
Thanks for reporting this!  I can reproduce this with an even smaller example:
```
import org.checkerframework.checker.nullness.qual.Nullable;

class Issue6804 {
   void method(String[] array, @Nullable Integer i) {
    String z = array[i];
  }
}
```
